/******************************************************************************
 * Copyright (c) ETSI 2017.
 *
 * This file is subject to copyrights owned by ETSI. Non-exclusive permission
 * is hereby granted, free of charge, to copy, reproduce and amend this file
 * under the following conditions: It is provided "as is", without warranty of any
 * kind, expressed or implied.
 * 
 * ETSI shall never be liable for any claim, damages, or other liability arising
 * from its use or inability of use.This permission does not apply to any documentation
 * associated with this file for which ETSI keeps all rights reserved. The present
 * copyright notice shall be included in all copies of whole or part of this
 * file and shall not imply any sub-license right.
 *
 * Modified by: Adrien Kirjak
 *
 ** @version  0.0.1
 ** @purpose  1:22.3.1, verify that the response and exception handling part cannot contain an altstep
 ** @verdict  pass reject
 *****************************************************************/

// The following requirements are tested:
// h) The use of else branches and the invocation of altsteps is not allowed.

module NegSyn_220301_CallOperation_002 {

	signature S();
	
	type port P procedure {
		inout S;
	} with {extension "internal"}
	
    type component GeneralComp {
		port P p;
	}
	
	function f_called() runs on GeneralComp {
		p.getcall(S:{});
		p.reply(S:{});
	}
	
	altstep a_handleReply() runs on GeneralComp {
		[] p.getreply {}
	}
	
    testcase TC_NegSyn_220301_CallOperation_002() runs on GeneralComp system GeneralComp {
        var GeneralComp v_ptc := GeneralComp.create;
		connect(self:p, v_ptc:p);
		v_ptc.start(f_called());
		p.call(S:{}) {
			[] p.getreply(S:{}) { setverdict(pass); }
			[] a_handleReply() {} // expected error
		}
    }

    control {
        execute(TC_NegSyn_220301_CallOperation_002(), 5.0);
    }
}
